x-pack/plugins/beats_management/server/routes/beats/enroll.ts

```diff
@@ -0,0 +1,151 @@
+export const createBeatEnrollmentRoute = (libs: CMServerLibs) => ({
+  method: 'POST',
+  path: '/api/beats/agent/{beatId}',
+  licenseRequired: REQUIRED_LICENSES,
+  config: {
+    auth: false,
+    validate: {
+      headers: Joi.object({
+        'kbn-beats-enrollment-token': Joi.string().required(),
+      }).options({
+        allowUnknown: true,
+      }),
+      payload: Joi.object({
+        host_name: Joi.string().required(),
+        name: Joi.string().required(),
+        type: Joi.string().required(),
+        version: Joi.string().required(),
+      }).required(),
+    },
+  },
+  handler: async (
+    request: FrameworkRequest
+  ): Promise<BaseReturnType | ReturnTypeCreate<string>> => {
+    const { beatId } = request.params;
+    const enrollmentToken = request.headers['kbn-beats-enrollment-token'];
+
+    const { status, accessToken } = await libs.beats.enrollBeat(
+      enrollmentToken,
+      beatId,
+      request.info.remoteAddress,
+      omit(request.payload, 'enrollment_token')
+    );
+
+    switch (status) {
+      case BeatEnrollmentStatus.ExpiredEnrollmentToken:
+        return {
+          error: { message: BeatEnrollmentStatus.ExpiredEnrollmentToken, code: 400 },
+          success: false,
+        };
+
+      case BeatEnrollmentStatus.InvalidEnrollmentToken:
+        return {
+          error: { message: BeatEnrollmentStatus.InvalidEnrollmentToken, code: 400 },
+          success: false,
+        };
+      case BeatEnrollmentStatus.Success:
+      default:
+        return {
+          item: accessToken,
+          action: 'created',
+          success: true,
+        };
+    }
+  },
+});
+*/
+
+export const registerBeatEnrollmentRoute = (router: IRouter) => {
+  // TODO: write to Kibana audit log file https://github.com/elastic/kibana/issues/26024
+  router.post(
+    {
+      path: '/api/beats/agent/{beatId}',
+      validate: {
+        params: schema.object({
+          beatId: schema.string(),
+        }),
+        body: schema.object(
+          {
+            host_name: schema.string(),
+            name: schema.string(),
+            type: schema.string(),
+            version: schema.string(),
+          },
+          { unknowns: 'ignore' }
+        ),
+      },
+      options: {
+        authRequired: false,
+      },
+    },
+    wrapRouteWithSecurity(
+      {
+        requiredLicense: REQUIRED_LICENSES,
+      },
+      async (context, request, response) => {
+        const beatsManagement = context.beatsManagement!;
+
+        const { beatId } = request.params;
+        const enrollmentToken = request.headers['kbn-beats-enrollment-token'] as string;
+        if (!enrollmentToken) {
+          return response.badRequest({
+            body: 'beats enrollment token required',
+          });
+        }
+
+        // TODO: fixme eventually, need to access `info.remoteAddress` from KibanaRequest.
+        const legacyRequest = ensureRawRequest(request);
```

Comment:
I need to access `legacyRequest.info.remoteAddress`, which is not exposed on `KibanaRequest`. I discussed with @restrry of the possibility to add it to core API. However as this plugin is the only one having this need and knowing that it will be removed for 8.0, I'm just using the private `ensureRawRequest` utility here.

I can implement the 'correct' way if that's preferable.

Reply:
Josh added one in https://github.com/elastic/kibana/pull/71019

Reply:
Ended up implementing it differently, so that PR no longer adds this API. I'm fine with this implementation for now and if other plugins (that aren't about to get deleted) also need this, we can re-evaluate adding it then.